lib/payment.spec.js

```diff
@@ -145,98 +103,136 @@ describe('Payment', () => {
-                expect(payment.recipient).to.eql(recipient);
+                expect(payment.recipient).to.eql(fixture.recipient);
             });
 
             it('does not have a valid signature', () => {
                 expect(payment.isSigned()).to.be.false;
             });
+
+            it('has an UUID as sender ref', function() {
+                expect(payment.senderRef).to.match(/^[a-f0-9]{8}-[a-f0-9]{4}-[a-f0-9]{4}-[a-f0-9]{4}-[a-f0-9]{12}$/i);
+            });
+
+            it('has sender data in JSON that is base64 encoding of sender ref', function() {
+                const expectedData = Buffer
+                    .from(JSON.stringify({ref: payment.senderRef}))
+                    .toString('base64');
+                expect(payment.toJSON().sender.data).to.eql(expectedData);
+            });
         });
 
         context('a signed Payment', () => {
-            let payment;
+            let payment, expectedSignedPayload;
 
             beforeEach(async () => {
-                payment = Payment.from(unsignedPayload, wallet);
+                payment = new Payment(MonetaryAmount.from(fixture.amount, fixture.currency.ct, fixture.currency.id), fixture.sender, fixture.recipient, wallet);
                 await payment.sign();
+                expectedSignedPayload = await fixture.createSignedPayment(fixture.createUnsignedPayment(payment.senderRef));
             });
 
             it('can be serialized to an object literal', () => {
-                expect(payment.toJSON()).to.eql(signedPayload);
+                expect(payment.toJSON()).to.eql(expectedSignedPayload);
             });
 
             it('can be registered with the API', () => {
                 payment.register();
-                expect(stubbedProvider.registerPayment).to.have.been.calledWith(signedPayload);
+                expect(stubbedProvider.registerPayment).to.have.been.calledWith(expectedSignedPayload);
             });
 
             it('has the supplied amount', () => {
-                expect(payment.amount).to.be.equalMoney(MonetaryAmount.from(amount, currency.ct, currency.id));
+                expect(payment.amount).to.be.equalMoney(MonetaryAmount.from(fixture.amount, fixture.currency.ct, fixture.currency.id));
             });
 
             it('has the supplied sender', () => {
-                expect(payment.sender).to.eql(sender);
+                expect(payment.sender).to.eql(fixture.sender);
             });
 
             it('has the supplied recipient', () => {
-                expect(payment.recipient).to.eql(recipient);
+                expect(payment.recipient).to.eql(fixture.recipient);
             });
 
             it('has a valid signature', () => {
                 expect(payment.isSigned()).to.be.true;
             });
+
+            it('has an UUID as sender ref', function() {
+                expect(payment.senderRef).to.match(/^[a-f0-9]{8}-[a-f0-9]{4}-[a-f0-9]{4}-[a-f0-9]{4}-[a-f0-9]{12}$/i);
```

Comment:
Maybe extract this regex as a separate variable, since it is used many times?

Reply:
Also, do you think it would be possible to parameterise the different test scenarios to avoid having to duplicate tests for each scenario? I have the impression you needed to add the same set of tests to a lot of different scenarios (for unsigned payment, signed payment etc)

Reply:
I was actually thinking of parameterizing, but decided it was too much work right now.